### PR TITLE
Refactor simulationmode client service (sorry for big PR 😬 )

### PR DESCRIPTION
**Pull request by KillariDev**
- Make Simulationmodeclient to be a set of functions instead of a class (fixes https://github.com/DarkFlorist/TheInterceptor/issues/222 )
- Fix couple race condition issues
- refactor prepend mode. fixes https://github.com/DarkFlorist/TheInterceptor/issues/221

### CODE REVIEW

File: extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts [hunk 1]
```diff
@@ -3,664 +3,634 @@ import { EthGetLogsResponse, EthereumUnsignedTransaction, EthereumSignedTransact
...
 	}
+	return tokenOwners
+}
+
+const getTokenBalancesAfter = async (
+	ethereumClientService: EthereumClientService,
+	signedTxs: EthereumSignedTransaction[] = [],
+	multicallResult: MulticallResponse,
+	blockNumber: bigint,
+) => {
```
Comment (MicahZoltu) [inline extension/app/ts/simulation/services/SimulationModeEthereumClientService.ts:623]:
```suggestion
const getTokenBalancesAfter = async (ethereumClientService: EthereumClientService, signedTxs: EthereumSignedTransaction[] = [], multicallResult: MulticallResponse, blockNumber: bigint) => {
```

**Review by MicahZoltu — APPROVED** [review]
Just some non-blocking style feedback.  Though, my eyes were glazing over by the end...

File: extension/app/ts/background/background.ts [hunk 2]
```diff
@@ -83,16 +85,24 @@ export function setEthereumNodeBlockPolling(enabled: boolean) {
...
+export async function refreshConfirmTransactionSimulation(
+	ethereumClientService: EthereumClientService,
+	simulationState: SimulationState, 
+	activeAddress: bigint,
+	simulationMode: boolean,
+	requestId: number,
+	transactionToSimulate: EthereumUnsignedTransaction,
+	website: Website,
+	userAddressBook: UserAddressBook
+) {
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/background/background.ts:97]: I think its cleaner when you see them side by side when there's so many of them

File: extension/app/ts/background/background.ts [hunk 3]
```diff
@@ -117,53 +127,37 @@ export async function refreshConfirmTransactionSimulation(activeAddress: bigint,
...
-async function handleSimulationMode(websiteTabConnections: WebsiteTabConnections, simulator: Simulator, socket: WebsiteSocket, website: Website, request: InterceptedRequest, settings: Settings): Promise<HandleSimulationModeReturnValue> {
+async function handleSimulationMode(
+	simulationState: SimulationState, 
+	websiteTabConnections: WebsiteTabConnections,
+	simulator: Simulator,
+	socket: WebsiteSocket,
+	website: Website,
+	request: InterceptedRequest,
+	settings: Settings
+): Promise<HandleSimulationModeReturnValue> {
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/background/background.ts:160]: I feel when there's a lot of parameters, its better to multiline them

File: extension/app/ts/background/background.ts [hunk 4]
```diff
@@ -291,73 +291,85 @@ async function handleSigningMode(simulator: Simulator, socket: WebsiteSocket, we
...
-		sendPopupMessageToOpenWindows({ method: 'popup_chain_update' })
-	}
+		if (updatedSettings.simulationMode) {
+			// update prepend mode as our active address has changed, so we need to be sure the rich modes money is sent to right address
+			const ethereumClientService = simulator.ethereum
+			await updateSimulationState(async () => {
+				const simulationState = (await getSimulationResults()).simulationState
+				const prependQueue = await getPrependTrasactions(ethereumClientService, updatedSettings, await getMakeMeRich())
+				return await setPrependTransactionsQueue(ethereumClientService, simulationState, prependQueue)
+			}, updatedSettings.activeSimulationAddress)
```
[the remark this answers is not in the document]
Comment (KillariDev) [reply extension/app/ts/background/background.ts:349]: I feel the most important parameters should generally be first, and here the update function is the most important one

**Pull request merged**